File: python/day1/day1.py
from typing import List
from functools import reduce
from itertools import combinations

def load_input_text() -> List[int]:
    with open("python/day-1-input.txt") as f:
        input = f.readlines()
    input = [int(i) for i in input]
    return input


def locate_matching_group(
    value_to_match: int,
    input_list: List[int],
    number_to_match: int
    ):
    combs = combinations(input_list, number_to_match)
    for c in combs:
        if reduce(lambda a,b:a+b, c) == value_to_match:
            return c


def reduce_by_multiply(input_list: List[int]) -> int:
    return reduce(lambda a,b: a*b, matches)


if __name__ == "__main__":
    i = load_input_text()
    matches = locate_matching_group(2020, i, 3)
    answer = reduce_by_multiply(matches)
    print(answer)

File: rust/aoc/src/day1.rs
use std::fs;
use itertools::Itertools;


pub fn load_input_text() -> Vec<i32> {
    // read the file and unwrap to panic if the file isn't readable
    let contents = fs::read_to_string("day-1-input.txt")
        .unwrap();

    let numbers: Vec<i32> = 
        contents.trim().split('\n')
        .map(String::from) // convert from reference &str to String value
        .map(|s| s.parse().unwrap()) // parse into i32 type (inferred)
        .collect();
        
    numbers
}


pub fn locate_matching_group( value_to_match: i32, input_list: Vec<i32>, number_to_match: i32) -> Vec<i32> {
    let combs = input_list
        .into_iter()
        .combinations(number_to_match as usize);
    
    for c in combs {
        if c.iter().fold(0, |a, &b| a + b) == value_to_match {
            return c
        }
    }
    panic!("Not found!")
}